### PR TITLE
MueLu & Tpetra: Half precision transfer ops

### PR DESCRIPTION
**Pull request by cgcgcg**
@trilinos/muelu @trilinos/tpetra 

## Motivation
- MueLu: Add option "transfers: half precision" (default: false). Converts transfers P and R to half precision.
- Tpetra: ETI of LocalCrsMatrixOperator for mixed scalar types

On my workstation I get these numbers for 100 SpMVs of a 1M Laplace2D matrix on 4 ranks:
```
======================================================================================================================================================

                                                        TimeMonitor results over 4 processors

Timer Name             MinOverProcs       MeanOverProcs      MaxOverProcs      MeanOverCallCounts    
------------------------------------------------------------------------------------------------------------------------------------------------------
MatVec original        0.2705 (1)         0.2707 (1)         0.2708 (1)        0.2707 (1)
MatVec low precision   0.2227 (1)         0.2228 (1)         0.2228 (1)        0.2228 (1)            
```

### CODE REVIEW

File: packages/muelu/src/Misc/MueLu_LowPrecisionFactory_decl.hpp [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+#include "MueLu_GraphBase.hpp"
+#include "MueLu_Level_fwd.hpp"
+#include "MueLu_SingleLevelFactoryBase.hpp"
+
+namespace MueLu {
+
+  /*!
+    @class LowPrecisionFactory class.
+    @brief Factory for building filtered matrices using filtered graphs.
```
Comment (jhux2) [inline packages/muelu/src/Misc/MueLu_LowPrecisionFactory_decl.hpp:62]: Copy and paste error?

File: packages/muelu/src/Misc/MueLu_LowPrecisionFactory_decl.hpp [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+    //@}
+
+    //! @name Build methods.
+    //@{
+
+    /*!
+      @brief Build method.
+
+      Builds filtered matrix and returns it in <tt>currentLevel</tt>.
```
Comment (jhux2) [inline packages/muelu/src/Misc/MueLu_LowPrecisionFactory_decl.hpp:97]: same as above.

File: packages/muelu/src/Misc/MueLu_LowPrecisionFactory_def.hpp [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+    const ParameterList& pL = GetParameterList();
+    std::string matrixKey = pL.get<std::string>("matrix key");
+
+    FactoryMonitor m(*this, "Converting " + matrixKey + " to half precision", currentLevel);
+
+    RCP<Matrix> A = Get< RCP<Matrix> >(currentLevel, matrixKey);
+
+    const bool isTpetra = A->getRowMap()->lib() == Xpetra::UseTpetra;
+    if (isTpetra && (false
+#if defined(HAVE_TPETRA_INST_DOUBLE) && defined(HAVE_TPETRA_INST_FLOAT)
```
Comment (jhux2) [inline packages/muelu/src/Misc/MueLu_LowPrecisionFactory_def.hpp:97]: oh, this is clever

**Review by jhux2 — APPROVED** [review]
yes!

File: packages/muelu/src/Misc/MueLu_LowPrecisionFactory_decl.hpp (hunk 1 above)
Comment (cgcgcg) [reply packages/muelu/src/Misc/MueLu_LowPrecisionFactory_decl.hpp]: Yup :-)

**Review by jhux2 — APPROVED** [review]
lgtm

**Review by jhux2 — DISMISSED** [review]
lgtm

File: packages/muelu/src/Misc/MueLu_LowPrecisionFactory_def.hpp (hunk 3 above)
Comment (cgcgcg) [reply packages/muelu/src/Misc/MueLu_LowPrecisionFactory_def.hpp:96]: Too clever for the Intel compiler...

**Review by jhux2 — APPROVED** [review]
lgtm

**Pull request merged**